### PR TITLE
Add deploy to gh-pages workflow

### PR DESCRIPTION
**Pull request by wryk**
Related to https://github.com/thomashoneyman/purescript-halogen-realworld/issues/21
Result on my fork available at https://wryk.github.io/purescript-halogen-realworld/

I kept workflow files simple resulting on lot of steps duplicated between ci and gh-pages (in comparison with single workflow and conditional jobs as a way to "share jobs between workflows").

I updated to Parcel 2 (currently in beta) to get the best bundle size (https://discourse.purescript.org/t/recommended-tooling-for-purescript-in-2020/1615/8). I expected [differential serving](https://v2.parceljs.org/getting-started/webapp/#differential-serving) to improve bundle size but the difference was negligible so I didn't used it. I kept previous cache folder git ignored so as not to provoke large number of non-tracked files on existing cloned repo.

Github Pages' "project sites" are served from a non-root path so I configured parcel to use relative links. This doesn't break current usage of `npm run bundle && npm run serve` on local as far is I know.

What do you think of those choices ?

### CODE REVIEW

File: .github/workflows/gh-pages.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+          zephyr: "0.3.2"
+
+      - name: Cache PureScript dependencies
+        uses: actions/cache@v2
+        with:
+          key: ${{ runner.os }}-spago-${{ hashFiles('**/*.dhall') }}
+          path: |
+            .spago
+            output
+      - uses: actions/setup-node@v1
```
Comment (thomashoneyman) [inline .github/workflows/gh-pages.yml]:
```suggestion

      - uses: actions/setup-node@v1
```

File: .github/workflows/gh-pages.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        run: npm ci
+
+      - name: Build source
+        run: npm run bundle
+
+      - name: Deploy to GitHub Pages
+        uses: peaceiris/actions-gh-pages@v3
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          publish_dir: ./dist
```
Comment (thomashoneyman) [inline .github/workflows/gh-pages.yml]:
```suggestion
          publish_dir: ./dist

```

File: .github/workflows/gh-pages.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
+name: GitHub Pages
```
Comment (thomashoneyman) [inline .github/workflows/gh-pages.yml]: This does have a bit of duplication with the main CI workflow. Ideally we could have a workflow with two jobs: a `build` job and a `deploy` job, which share the results of building the source and the second just bundles and publishes. On the other hand, I know that it's super easy to get bogged down in CI so I don't want to force you to do that. So I'm going to accept this as-is, but if you would like to try merging them together then I welcome that as a further addition.

**Review by thomashoneyman — APPROVED** [review]

**Comment by thomashoneyman** [comment]
I'd like to leave this open until tomorrow so you have a chance to weigh in @wryk if you have anything to add, but otherwise I'll go ahead and merge this!

**Comment by wryk** [comment]
Steps duplication bother me too, I'll take a deeper look at Github Actions to improve that. Don't merge this @thomashoneyman, I'll ping you back when its done.

**Comment by wryk** [comment]
I've pushed changes that fix steps duplication, can you take a look a them @thomashoneyman ?

- push on main https://github.com/wryk/purescript-halogen-realworld/runs/1003850091?check_suite_focus=true
- pull request on main (skip deploy) https://github.com/wryk/purescript-halogen-realworld/runs/1003906293?check_suite_focus=true

I initially naively organized workflow in multiple jobs (setup, build, test, deploy) but I learned jobs doesn't use the same runner (they can run in parallel I don't know why I expected them to run on the same runner). [Jobs outputs](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjobs_idoutputs) and [upload/download actions](https://docs.github.com/en/actions/configuring-and-managing-workflows/persisting-workflow-data-using-artifacts#passing-data-between-jobs-in-a-workflow) doesn't seems adapted to share tooling like Spago and NPM between jobs.

I could have done only two jobs (build, deploy) and share bundle between them with upload/download actions but I didn't feel like it was worth storing artifacts only for one job.

**Comment by thomashoneyman** [comment]
You can have a job depend on an earlier job, which lets you make them sequential. Also you’d need to run setup-purescript & setup-node each job (though the installed tools are cached).

But I agree in this case it’s not worth persisting artifacts between runs — the simple conditional looks good to me!

**Comment by wryk** [comment]
That was how I organized the workflow before realizing successive jobs will not have purescript and node without running setup steps again in them.

```yaml
jobs:
  setup:
    runs-on: ubuntu-latest

    steps:
      # checkout, setup-purescript, cache purescript, setup-node, cache node, install npm deps

  build:
    needs: setup
    runs-on: ubuntu-latest

    steps:
      # build source

  test:
    needs: build
    runs-on: ubuntu-latest

    steps:
      # run tests

  deploy:
    needs: test
    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
    runs-on: ubuntu-latest

    steps:
      # run bundle, deploy to gh pages
```

With steps duplication fixed I'm no longer bothered by my pull request. Do you want any other changes ?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -7,10 +7,11 @@ on:
     branches: [main]
 
 jobs:
-  build:
+  main:
```
Comment (thomashoneyman) [inline .github/workflows/ci.yml]: Do you mind leaving this job name as `build`?

**Comment by thomashoneyman** [comment]
Other than renaming the job I'm happy to go ahead and merge this.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (wryk) [reply .github/workflows/ci.yml]: Yes sure

**Review by thomashoneyman — APPROVED** [review]

**Pull request merged**